[#14] Implement proposals and voting logic with tests

rinn7e opened the pull request:
Problem: After #5 and #6 we will have FA2 entrypoints, but there
are more in the specification. We are going to focus implementing
proposals and voting in this issue.

Solution: Add all entrypoints related to proposals and voting. Add
tests for these entrypoints.

## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

Resolves #14 

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

rinn7e commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -5,72 +5,208 @@
+  , sProposalsIdSortByDate :: [ProposalId] -- Needed for `flush`
+
+  -- Configuration (TODO: make it a separate record)
+  , sRejectedProposalReturnValue :: Natural
```
Took me a while to figure out how to make the types that contain `proposalMetadata` type params work (Storage, Proposal etc ...)

This is still pretty much WIP, only `propose` entrypoint is implemented which is still incomplete.

There seems to be some changes from the spec feedback that will change most of the things here as well. 

So the MR is not yet ready for review, submitted for showing current progress only. I probably focus on adding tests for these entrypoints first instead until we got the first initial feedback for the spec.

rinn7e replied:
@gromakovsky I've updated this MR. Even though it is still WIP, I would like some of your feedback whether or not I'm on the right track, Ivan.

I have implemented:
- `propose` entrypoint: It can now check if a proposal is valid, and add to the storage. 
  - It is still in completed because it relies on `creditTo` and `debitFrom`
- `propose` tests: Current it can check for some invalid proposals.

Incompleted:
- `cProposalCheck` is current just a `Natural` which stand for the minimum amount of token required to create a proposal. I will wait for the feedback before making this a lambda.
- `vote` and other entrypoints/tests are not yet implemented.

rinn7e replied:
I just figure this out. I have update this by adding the `Config`. Let me know if this is good enough, Ivan.

gromakovsky replied:
Yes, looks good now.

rinn7e replied:
Thanks, Ivan. I'll continue to implement the rest of entrypoints and adding the tests. I'll let you know when this is ready for review again.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,106 @@
+  => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+checkSenderUnfrozenToken = do
+  duupX @2
+  push (0 :: Natural) -- unfrozen token_id
```
It'd be nice to have `unfrozenTokenId :: TokenId` defined somewhere. Same for `frozenTokenId`.

rinn7e replied:
I have added this in `BaseDAO.Types`.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -14,78 +19,203 @@ module Lorentz.Contracts.BaseDAO.Types
-instance ParameterHasEntrypoints Parameter where
-  type ParameterEntrypointsDerivation Parameter = EpdDelegate
+instance (HasAnnotation pm, IsoValue pm) => ParameterHasEntrypoints (Parameter pm) where
+  type ParameterEntrypointsDerivation (Parameter pm) = EpdRecursive
```
Why not keep `EpdDelegate`?

rinn7e replied:
Ater adding the type params `proposalMetadata`, EpdDelegate no longer works:
```
    • Could not deduce: (Michelson.Typed.Scope.FailOnOperationFound
                           (Michelson.Typed.Scope.ContainsOp (ToT pm)),
                         Michelson.Typed.Scope.FailOnNestedBigMapsFound
                           (Michelson.Typed.Scope.ContainsNestedBigMaps (ToT pm)))
        arising from the superclasses of an instance declaration
      from the context: (HasAnnotation pm, IsoValue pm)
        bound by the instance declaration
        at src/Lorentz/Contracts/BaseDAO/Types.hs:184:10-82
    • In the instance declaration for
        ‘ParameterHasEntrypoints (Parameter pm)’
    |       
184 | instance (HasAnnotation pm, IsoValue pm) => ParameterHasEntrypoints (Parameter pm) where
```

rinn7e replied:
I've updated this, and all other comments as well, Ivan.

rinn7e commented on test/Test/BaseDAO/Nettest/FA2.hs:
```diff
@@ -1,61 +1,26 @@
 -- SPDX-License-Identifier: LicenseRef-MIT-TQ
 
-module Test.BaseDAO.FA2
+module Test.BaseDAO.Nettest.FA2
```
I remove the `Nettest` and reorder some of the test files structure.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -10,30 +10,23 @@ import Lorentz
+  :: forall store pm. (StorageC store pm)
+  => Config pm -> '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+checkIsProposalValid Config{..} = do
+  duupX @1
```
Isn't it just `dup`?

gromakovsky commented on test/Test/BaseDAO/FA2.hs:
```diff
@@ -126,7 +127,7 @@ validTransferOwnerScenario = uncapsNettest $ do
 updatingOperatorScenario :: (Monad m) => NettestImpl m -> m ()
 updatingOperatorScenario = uncapsNettest $ do
-  ((owner1, _), (owner2, _), dao, _) <- originateBaseDAO
+  ((owner1, _), (owner2, _), dao, _) <- originateBaseDAO DAO.defaultConfig
```
`originateBaseDAO DAO.defaultConfig` seems to be used in many places, I would define a helper for that. Or hardcode `defaultConfig` in `originateBaseDAO` if we don't use anything else in tests.

rinn7e commented:
@gromakovsky I've added `proposalMetadata`, `setVotingPeriod`, `setQuorumThreshold` and `vote`.

A few assumptions, I've made which are:
- `setVotingPeriod`, `setQuorumThreshold`: will affect all ongoing proposals since all proposals use the same `votingPeriod` and `quorumThreshold` stored in `Storage`. I'm not sure if this is what we prefer. We can avoid affecting ongoing proposals' by storing their specific votingPeriod and quorum inside the `Proposal` itself, so that is what we prefer. Hope to get your advice on this.

- Currently, the votingPeriod is the day counts in: `Natural`, and I assume that it starts right away when a proposal passed the `cProposalCheck`.

- Since `vote` accepts a list of `VoteParam`, if one of the `VoteParam` failed, the ones that are succeeded are still recorded. I'm not exactly sure how to do this atomically (if one failed, the all should fail). Hope to get your advice on this as well.

I will work on `flush` next, and implements the tests for these entrypoints.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -4,10 +4,50 @@
+ensureVotingPeriodIsNotOver = do
+  toField #pStartDate
+  swap
+  stToField #sVotingPeriod; int;
```
There are some redundant `;`s in many places. Trailing `;` is redundant.

rinn7e replied:
I have removed the unnecessary `;`.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO.hs:
```diff
@@ -33,7 +33,11 @@ baseDaoContract config@Config{..} = defaultContract $ contractName cDaoName $ do
-    -- , #cVote /-> vote --TODO
+    , #cProposalMetadata /-> proposalMetadata
+    , #cVote /-> vote
+    , #cSetVotingPeriod /-> setVotingPeriod
```
AFAIU this entrypoint's name will be `setVotingPeriod`, however, our specification says it should be `set_voting_period`. Same applies to some other entrypoints.

rinn7e replied:
Got this, Ivan. I've fixed this accordingly.

gromakovsky commented:
> setVotingPeriod, setQuorumThreshold: will affect all ongoing proposals since all proposals use the same votingPeriod and quorumThreshold stored in Storage.

I think it's fine, but we should document such details in the specification, so that they are clear. We can revisit it later.

> Currently, the votingPeriod is the day counts in: `Natural`, and I assume that it starts right away when a proposal passed the cProposalCheck.

I think a day is too much for granularity here. Even though I agree that usually voting will usually last several days, I can imagine that someone may want to conduct quick voting. Since we are developing something very general, I think we shouldn't limit voting periods to be multiples of a day. I would specify it in seconds. This way we'll be able to have tests for this behavior and run them in a real network. If it's at least a day or an hour, running such tests will be unrealistic.

gromakovsky commented:
>  I'm not exactly sure how to do this atomically (if one failed, the all should fail).

AFAIU it already works this way. If one vote fails, we do `FAILWITH` which stops contract execution and doesn't mutate any state.

rinn7e commented:
> I think it's fine, but we should document such details in the specification, so that they are clear. We can revisit it later.

Should I update the spec in this MR, or the currently opened "Update Spec" MR?

gromakovsky commented:
> Should I update the spec in this MR, or the currently opened "Update Spec" MR?

Hm, I think let's do it in "Update Spec", because it's close to be finished and probably we'll merge it soon, while this one still lacks some things which will likely take more time.

rinn7e commented:
@gromakovsky I have added `flush` entrypoint, and fixed a few issues with previous entrypoints. I have added the tests for those entrypoints as well but they are incomplete.

One note is that, I need to add `pVoters :: [(Address, Natural)]` in `Proposal` since we need to keep track of voter address and the amount of token he has spent so that we can distribute it back when the proposal is `flush`. Not sure if storing in a List is a good idea, probably `Set` is a better choice in this case.

Next week, I will try various way to improve above issues, handle the TODOs that are left, and add more tests.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -3,90 +3,332 @@
+  , sQuorumThreshold :: QuorumThreshold
+
+  , sProposals :: BigMap ProposalId (Proposal proposalMetadata)
+  , sProposalsIdSortByDate :: [ProposalId] -- Newest first (probably change to set)
```
Having unbounded lists in the storage is dangerous. For example, if there are too many proposals, the contract may become unusable because storage deserialization will become too expensive. So we should explicitly limits lengths of all strict data types. For example, we can require that there is no more than 500 active proposals and it's impossible for one proposal to have more than 1000 votes. I think it can be part of compile-time configuration (I mean these numbers: 500 and 1000). Then we should mention it in the specification.

If there are other unbounded types, we should limit them as well. For example, `QuorumThreshold` is `Natural` and `nat` type is unbounded, so it can be arbitrarily large as well. Even though `sQuorumThreshold` is available only to the admin who we trust, I think we should limit it as a safety measure.

rinn7e replied:
I have added the bounded value in `Config` as below: 

```haskell
data Config = Config pm 
  { -- Bounded
  , cMaxProposals :: Natural
  , cMaxVotes :: Natural -- including both upvotes and downvotes
  , cMaxQuorumThreshold :: Natural -- Must be smaller than or equal to MaxVotes or else the threshold could be never met.
  , cMinQuorumThreshold :: Natural
  , cMaxVotingPeriod :: Natural -- Maximum of seconds allow to be set
  , cMinVotingPeriod :: Natural
  }

defaultConfig :: Config pm
defaultConfig =
  { 
  , cMaxVotingPeriod = 60 * 60 * 24 * 30 -- 1 month
  , cMinVotingPeriod = 60 * 60 -- 1 hour

  , cMaxQuorumThreshold = 1000
  , cMinQuorumThreshold = 1

  , cMaxVotes = 1000

  , cMaxProposals = 500
  }
```

I also added a check using the abouve bounded-value in `propose`, `vote`, `setQuorumThreshold` and `setVotingPeriod` entrypoints as well. Let me know if this is good enough or not, Ivan.

gromakovsky replied:
There new checks look ok. But there is still some work that needs to be done in this PR. Please ping me when it's ready.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -86,6 +86,22 @@ checkProposerUnfrozenToken = do
+    failCustom_ #mAX_PROPOSALS_REACHED
+  else nop
+
 -- TODO: implement
```
The PR where it's implemented got merged recently, so it can be implemented/adapted here now.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -86,6 +86,22 @@ checkProposerUnfrozenToken = do
+  stToField #sProposalsIdSortByDate
+  size
+  push (1 :: Natural)
+  add
```
Can we avoid adding 1 by replacing `<.` check with `<=.` below?

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -265,21 +280,39 @@ submitVote = do
+  dip $ toField #vVoteAmount
+  add; toNamed #newAmount
+
+  push (cMaxVotes); toNamed #maxAmount
```
Brackets look redundant.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -265,21 +280,39 @@ submitVote = do
+
 -- | Vote
 -- TODO: If one of the vote param fails, does it cancel all the votes?
 -- Current solution, will fail if one of the vote params causes error.
```
I think it's ok.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -3,89 +3,351 @@
+  -- only 50 tokens will be unfrozen and other 50 tokens will be burnt.
+
+  , cDecisionLambda :: forall s
+      . (Proposal proposalMetadata) & (Storage proposalMetadata) & s
```
`:` should be better than `&` for the reason described in https://issues.serokell.io/issue/TM-307. Moreover, according to that issue `&` may end up being deprecated. So I think it's better to use `:` in new code.

rinn7e replied:
@gromakovsky I've updated this, add flush tests, as well as handle all the left TODOs.

2 issues that I found which I marked as `[issue 1]` and `[issue 2]` in the source code.

- [issue 1]: Since it is currently not possible to do a `delay 3 seconds` in Nettest, we cannot test if a proposal voting period is over. I have a workaround for the tests by having a `cForceVotingPeriod` in the `Config` which is just to ensured that the tests work.
  - This issue is linked with issue in morley: https://gitlab.com/morley-framework/morley/-/issues/351
  - We will remove the `cForceVotingPeriod` after the above issue is resolved. I propose we handle this in a separate issue.

- [issue 2]: It is currently not possible to check just field value in `Storage` via Nettest. This is related to `Storage` containing `BigMap`. This is the related issue in morley: https://gitlab.com/morley-framework/morley/-/issues/415#note_435327096
  - I propose we handle this in a separate issue as well.

Some other problems that I have is that, some implementations are depended on `burn` entrypoints, so I add a `TODO: [#15]` which should be handled in #15 itself. (Example: slash value need to be burn`.

The last thing is testing `cDecisionLambda`, I'm not really sure how to test this.

For the rests, everything is ready for review.

Hope to get your advice on this, Ivan

gromakovsky replied:
> [issue 1]: 

I wrote [a comment](https://serokell.slack.com/archives/CF2UVU8RX/p1603736318499400?thread_ts=1603706879.484000&cid=CF2UVU8RX) about that. I think it can be implemented quite easily for Emulator, so I would wait for that.

> [issue 2]:

AFAIu a more relevant upstream issue is https://gitlab.com/morley-framework/morley/-/issues/346. Anyway, I don't mind handling it in a separate issue since the upstream issue is not started yet.

> slash value need to be `burn`

Hm, AFAIU, `burn` is basically `debitFrom`, can't you use `debitFrom`?

> The last thing is testing cDecisionLambda, I'm not really sure how to test this.

Well, you can define a decision lambda that does _something_, for example it can transfer 100 mutez to a predefined address when proposal passes. And then check that 100 mutez is actually transferred.

rinn7e replied:
I've created 2 separate issues to track the above problem.

I also add `burnFrozenToken` to make the slash function works properly now. 
I added a test for `cDecisionLambda` as well.

Related to using `ByteString` as key, I've refactored that as well. You can check it, Ivan.

gromakovsky commented:
There are many conflicts now:
```
docs/specification.md
src/Lorentz/Contracts/BaseDAO.hs
src/Lorentz/Contracts/BaseDAO/FA2.hs
src/Lorentz/Contracts/BaseDAO/Types.hs
test/Test/BaseDAO/FA2.hs
test/Test/Management.hs 
```

gromakovsky commented on test/Test/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,381 @@
+      ]
+  , testGroup "Other:"
+      [ testCase "can call proposalMetadata" $
+          ignore $ nettestTestExpectation proposalMetadata
```
Why is it ignored now?

rinn7e replied:
I've fixed this.

gromakovsky commented on test/Test/Common.hs:
```diff
@@ -2,15 +2,30 @@
+import qualified Lorentz.Contracts.BaseDAO as DAO
+
+
+ignore :: Expectation -> Expectation
```
Is it the only reason why `hspec` dependency was added? You can change `Expectation` to `Assertion` to avoid adding an additional dependency.

rinn7e replied:
I've removed this for now, since we do not use it.

gromakovsky commented on test/Test/BaseDAO/ProposalConfig.hs:
```diff
@@ -0,0 +1,69 @@
+      toField #pProposer
+      dip $ do
+        push (10 :: Natural)
+        push DAO.unfrozenTokenId; pair
```
Isn't it simpler to `push (10 :: Natural, DAO.unfrozenTokenId)` here?

rinn7e replied:
Updated this.

gromakovsky commented on test/Test/BaseDAO/ProposalConfig.hs:
```diff
@@ -0,0 +1,69 @@
+      ediv
+      ifSome
+        (do
+          unpair; dip drop
```
AFAIU `unpair; dip drop` is just `car`.

rinn7e replied:
Fixed this.

gromakovsky commented:
The specification should be updated with regards to proposal IDs. Currently there is this
> Proposals are identified by a natural number. A counter is used to assign identifiers to proposals. So proposal IDs are 0, 1, 2…

which is not true anymore.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -40,73 +153,201 @@ instance TypeHasDoc MigrationStatus where
+
+type ProposalKey = ByteString
+
+data Proposal proposalMetadata = Proposal
```
It'd be nice to comment more types added in this module.

rinn7e replied:
I added a comment to describe this type as well as update the spec.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -40,73 +153,201 @@ instance TypeHasDoc MigrationStatus where
+type VotingPeriod = Natural
+
+-- | QuorumThreshold that a proposal need to meet
+-- quorum_threshold = upvote - downvote
```
I see at least 2 possible interpretations of quorum threshold. It can be `positive + negative` or `positive - negative`. I guess it can also be just `positive`. I am not sure which one is correct, I've asked [here]((https://tocquevillegroup.slack.com/archives/C01BTEPMY5U/p1603814934010800).

gromakovsky replied:
UPD: so it's actually the total number of votes.

rinn7e replied:
I have refactored this according to our discussion from slack. Let me know if this is good enough, Ivan.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -172,3 +413,72 @@ instance CustomErrorHasDoc "fORBIDDEN_XTZ" where
   customErrClass = ErrClassActionException
   customErrDocMdCause =
     "Received some XTZ as part of a contract call, which is forbidden"
+type instance ErrorArg "fAIL_PROPOSAL_CHECK" = ()
```
```suggestion

type instance ErrorArg "fAIL_PROPOSAL_CHECK" = ()
```

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,682 @@
+  stToField #sVotingPeriod; int
+  -- push (60 * 60 * 24 :: Natural); -- one day
+  -- mul; -- multiply with the voting period
+  add; -- add to the start_date
```
AFAIU commented out code above can be removed now. Also `;` looks redundant here.

rinn7e replied:
Fixed this.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,682 @@
+  => ProposeParams pm : store : s :-> ProposeParams pm : store : s
+ensureProposalIsUnique = do
+  duupX @2; duupX @2
+  sender; swap; pair; pack @(ProposeParams pm, Address)
```
1. I propose to define a type alias for data that is packed, otherwise it's easy to pass something wrong to `pack`. I mean, if you pass `(Address, ProposeParams)` instead of `(ProposeParams, Address)`, it will still compile. I would even go further and define something like `toProposalId` instruction and not use `pack` directly (outside of `toProposalId`).
2. I previously suggested that we also `blake2b` packed data, so that proposal IDs look more uniform and don't depend on proposal size. For large a proposal its ID can be too long.

rinn7e replied:
I've created a helper function `toProposalKey`  as well as add `blake2b` after `pack`.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,682 @@
+-- incrementProposalId
+--   :: forall store pm. (StorageC store pm)
+--   => '[ProposeParams pm, store] :-> '[ProposeParams pm, store]
+-- incrementProposalId = do
```
Probably this thing can be removed now as well.

rinn7e replied:
Removed

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,682 @@
+          sub; isNat
+          ifSome
+            (do
+              nop
```
I think `nop` is redundant and then you don't need `do`.

rinn7e replied:
I've refactored this.

gromakovsky commented on test/Test/BaseDAO/ProposalConfig.hs:
```diff
@@ -34,12 +34,8 @@ config = DAO.defaultConfig
-          push (0 :: Natural)
-        )
+      ifSome car $ do
+        push (0 :: Natural)
```
`do` here looks redundant, I think it can be written as one line.

rinn7e replied:
Fixed this.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,674 @@
+  stackType @[Proposal pm, store, ProposeParams pm]
+  some
+  duupX @3; sender; swap; pair; toProposalKey
+  -- stToField #sProposalIdCounter
```
Should be removed.

rinn7e replied:
I've removed this.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,674 @@
+vote config = do
+  iter $ do
+    duupX @2
+    duupX @2
```
Minor: you could use `dupTop2` here, it should be available in the `lorentz` version we use here (if it's not, ignore my comment).

rinn7e replied:
`dupTop2` is indeed available. I've replaced all `duupX @2; duupX @2` with `dupTop2` instead.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,674 @@
+    cons
+    stSetField #sProposalKeyListSortByDate
+
+-- | Update quroum threshold of all ongoing and new proposals.
```
Is this description correct? Quorum threshold is one for the whole contract and is updated by another entrypoint. I think this function can be better described.

rinn7e replied:
This is wrong comment for `flush`. I removed this, and add proper `flush` description instead.

I also made some changes to `setQuorumTheshold` comment as well.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -40,73 +155,209 @@ instance TypeHasDoc MigrationStatus where
+  , pProposer :: Address
+  , pProposerFrozenToken :: Natural
+
+  , pVoters :: [(Address, Natural)] -- This can be redundant with "Upvote/Downvote"
```
> -- This can be redundant with "Upvote/Downvote"

What do you mean? AFAIU it's not redundant because we need to unfreeze tokens.

rinn7e replied:
That comment was outdated, I removed this.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,674 @@
+            push False
+
+
+-- Used in "flush". When unfreezing the tokoens of voters/proposers
```
```suggestion
-- Used in "flush". When unfreezing the tokens of voters/proposers
```

rinn7e replied:
Fixed.

gromakovsky commented on src/Lorentz/Contracts/BaseDAO/Proposal.hs:
```diff
@@ -0,0 +1,674 @@
+  swap; dip swap
+  stackType @[Address, store, "unfreeze_value" :! Natural, Proposal pm, ProposalKey, [Operation]]
+  duupX @2; duupX @2
+  push (frozenTokenId); swap; pair
```
```suggestion
  push frozenTokenId; swap; pair
```

rinn7e replied:
Fixed.

PruStephan commented on src/Lorentz/Contracts/BaseDAO/Types.hs:
```diff
@@ -40,73 +155,209 @@ instance TypeHasDoc MigrationStatus where
+
+data VoteParam = VoteParam
+  { vProposalKey :: ProposalKey
+  , vVoteType :: Bool
```
Why don't you use `VoteType` here?

rinn7e replied:
Yeah, I forgot. Fixed this.

PruStephan commented on src/Lorentz/Contracts/BaseDAO/FA2.hs:
```diff
@@ -90,8 +92,8 @@ transfer = do
-      :: forall f. Address & Storage & f
-      :-> Address & Storage & f
+      :: forall f. Address : Storage pm : f
+      :-> Address : Storage pm : f
```
What is the difference between `:` and `&`, since nothing else had changed

rinn7e replied:
I believe `&` will be deprecated. It is recommended to use `:`, from what I see, they serve the same purpose.

PruStephan reviewed (commented):
LGTM, Except for two minor questions, that I just want to clear to myself

gromakovsky reviewed (approved):
Lorentz code is hard to review due to its low-levelness, but overall it looks fine. I think we can proceed with merge and refine later if something appears to be wrong.
